vm-virtio/src/transport/pci_device.rs

```diff
@@ -599,3 +630,83 @@ impl BusDevice for VirtioPciDevice {
         self.configuration.read_config_register(reg_idx)
     }
 }
+
+impl Device for VirtioPciDevice {
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn read(&mut self, addr: GuestAddress, data: &mut [u8], _io_type: IoType) {
+        self.read_bar(addr.0 - self.bar_base_addr, data)
+    }
+
+    fn write(&mut self, addr: GuestAddress, data: &[u8], _io_type: IoType) {
+        self.write_bar(addr.0 - self.bar_base_addr, data)
+    }
+
+    fn set_resources(&mut self, res: &[IoResource], irq: Option<IrqResource>) {
+        // For now, virtio device only designs one BAR.
+        if let Some(r) = res.iter().next() {
+            if let Some(a) = r.addr {
+                let config = PciBarConfiguration::default()
+                    .set_register_index(0)
+                    .set_address(a.0)
+                    .set_size(r.size as u64);
+
+                self.bar_base_addr = a.0;
+        
+                let virtio_pci_bar =
+                    self.configuration.add_pci_bar(&config).map_err(|e| {
+                    PciDeviceError::IoRegistrationFailed(a.0, e)
+                    }).unwrap() as u8;
+
+                // Once the BARs are allocated, the capabilities can be added to the PCI configuration.
+                if self.add_pci_capabilities(virtio_pci_bar).is_err() {
+                    println!("pci capability adding error!");
+                }
+            }
+        } else {
+            // BAR allocated failed.
+            return;
+        }
+
+        for r in res[1..].iter() {
+            // Set the device specific BARs allocated resource back.
+            for config in self.device.get_device_bars() {
+                if let Some(a) = r.addr {
+                    config.set_address(a.0);
+                    let _device_bar = self.configuration.add_pci_bar(&config).map_err(|e| {
+                        PciDeviceError::IoRegistrationFailed(a.0, e)
+                    });
+                } else {
+                    // BAR allocated failed.
+                    return;
+                }
+            }
+        }
+        // IRQ
+        match irq {
+            None => return,
+            Some(IrqResource(irq_num)) => {
+                // Assign IRQ to the virtio-pci device
+                let irqfd = EventFd::new(EFD_NONBLOCK).ok();
+
+                match irqfd {
```

Comment:
@sameo this is the problem place.

Reply:
@sboeuf Could you also take a look at this? Since the irqfd is moved for composing callback in msix patch.

Reply:
@sboeuf The issue is, when we add_virtio_pci_device(), legacy irq case, we need 1. device_manager::register_device, to allocate resource, bars, etc, 2. reuse the EventFd, created when device::set_resources, and register_irqfd with the allocated irq_num. So there would be problem for the second step, because we move the irqfd to callback...

Reply:
Can't you clone the EventFd? Or use the RawFd from `.as_raw_fd()` function?